### DISCUSSION TITLE
How to implement "find nearest user in your area"?

### DISCUSSION THREAD

**Discussion by tananga** [appwrite/appwrite#3257] [General]
Hello. I want to do app like tinder where users can find other users base on their location. It would be very easy in graph database but I dont know how to implement it in AppWrite. Do you have any suggestions? Requests should be as cheap as possible 😁

**Comment by gewenyu99** [comment]
:wave: Hi there!

This is related to: https://github.com/appwrite/appwrite/issues/388
But I will answer it here, too.

Checkout this article: https://levelup.gitconnected.com/building-a-location-based-app-with-appwrite-48a2e2b6d4c2

If you can settle with a rectangular search boundary, you should be able to build something similar. I wouldn't call the request cheap, though. I would suggest you chime in on the issue mentioned above and request proper geolocation features to be implemented in our DB.

For the moment, I hope the article gives you some inspiration for a work around.

Thanks!

**Comment by Meldiron** [comment]
A Few months ago I suggested a solution using Geohashes on Discord. Copying message here:

---

Image I have in my head for implementation:
`restaurants` collection with info about the place, like `name`, `openHours`, `phoneNumber`
`restaurantsGeohashes` with many-to-many relationship, having `restaurantId` and `geohash`. Here we will need index on `geohash`.

Lets say I add restaurant `BestPizza` on location `u2fkffcfy`, then I add one document into `restaurants`, and 9 documents into `restaurantsGeohashes`:

```
# restaurants
{ name: 'BestPizza', openHours: '08-17', 'phoneNumber': '+420204958...' }

# Let's say newly created restaurant $id is 'abcd1234'
# restaurantsGeohashes
{ restaurantId: 'abcd1234', geohash: 'u2fkffcfy' },
{ restaurantId: 'abcd1234', geohash: 'u2fkffcf' },
{ restaurantId: 'abcd1234', geohash: 'u2fkffc' },
{ restaurantId: 'abcd1234', geohash: 'u2fkff' },
{ restaurantId: 'abcd1234', geohash: 'u2fkf' },
{ restaurantId: 'abcd1234', geohash: 'u2fk' },
{ restaurantId: 'abcd1234', geohash: 'u2f' },
{ restaurantId: 'abcd1234', geohash: 'u2' },
{ restaurantId: 'abcd1234', geohash: 'u' },
```

Yes, writing will take some time, but it's fine.. Appwrite itself is optimised for reads, not writes, thanks to indexes.

Now if you want to get list of restaurants in your location by your geohash, you can simply:
```
appwrite.database.listDocuments('restaurantsGeohashes', [ Query.equal('geohash', 'YOUR_GEOHASH') ]);
```

You can also use some client-side library (dart_geohash for example) that can take geohash (or lat&lng) and radius as input, an returns array of all geohashes that would fit your search query. Then you can use Appwrite to query all of them:
```
appwrite.database.listDocuments('restaurantsGeohashes', [ Query.equal('geohash', ['GEOHASH1', 'GEOHASH2', 'GEOHASH3', 'GEOHASH4']) ]);
```

Oh and last step, you will only have array of restaurant IDs.. So finally you will query original collection to get details about paces you are looking for:

```
appwrite.database.listDocuments('restaurants', [ Query.equal('$id', [ 'ID1', 'ID2', 'ID3', 'ID4' ]) ]);
```

---

Regarding Geohash itself, look it up, there are a bunch of libraries. It allows you to express a location (lat&lng) using a string that can be easily searchable.

**Comment by tananga** [comment]
Thank you for answers. The lack of geolocation is painful but I hope that in the future you will implement it. Now you should focus on AppWrite Cloud. I will try @gewenyu99 version first :)
